### PR TITLE
Document changes in the deprecation error handler

### PR DESCRIPTION
**Pull request by llaakkkk**
See https://github.com/symfony/symfony/pull/29211
See https://github.com/symfony/symfony/issues/28048

Closes #10701

### CODE REVIEW

File: components/phpunit_bridge.rst [hunk 1]
```diff
@@ -30,6 +30,8 @@ Installation
 
     $ composer require --dev "symfony/phpunit-bridge:*"
 
+Alternatively, you can clone the `<https://github.com/symfony/phpunit-bridge>`_ repository.
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: Thai should be removed

File: components/phpunit_bridge.rst [hunk 2]
```diff
@@ -30,6 +30,8 @@ Installation
 
     $ composer require --dev "symfony/phpunit-bridge:*"
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: Do we want the „:*“ ?

File: components/phpunit_bridge.rst [hunk 3]
```diff
@@ -246,6 +298,9 @@ class autoloading time. This can be disabled with the ``debug-class-loader`` opt
         </listener>
     </listeners>
 
+.. versionadded:: 4.2
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: Please add a blank line afterwards

File: components/phpunit_bridge.rst (hunk 2 above)
Comment (wouterj) [reply components/phpunit_bridge.rst]: I would remove the start, relying on `composer require` to pick a good version constraint for us

Comment (llaakkkk) [reply components/phpunit_bridge.rst]: @OskarStark @wouterj so I don't need to remove the star for now?

Comment (wouterj) [reply components/phpunit_bridge.rst]: Please remove the star. Using a start would result in `"symfony/phpunit-bridge": "*"` being added to your `composer.json`, which is a bit scary. Without the star, the command figures out the version constraint so it'll end up as `^4.2` or so in the `composer.json`, which is better imo

**Review by OskarStark — APPROVED** [review]
👌🏻💥

File: bundles/best_practices.rst [hunk 4]
```diff
@@ -197,9 +197,9 @@ of Symfony and the latest beta release:
         include:
               # Minimum supported dependencies with the latest and oldest PHP version
             - php: 7.2
-              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="weak_vendors"
+              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="max[internal]=0"
```
Comment (greg0ire) [inline bundles/best_practices.rst]: `internal` has been renamed to `self`
```suggestion
              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="max[self]=0"
```

File: bundles/best_practices.rst [hunk 5]
```diff
@@ -197,9 +197,9 @@ of Symfony and the latest beta release:
...
             - php: 7.0
-              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="weak_vendors"
+              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="max[internal]=0"
```
Comment (greg0ire) [inline bundles/best_practices.rst]:
```suggestion
              env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest" SYMFONY_DEPRECATIONS_HELPER="max[self]=0"
```

File: components/phpunit_bridge.rst [hunk 6]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
-will often use ``SYMFONY_DEPRECATIONS_HELPER=weak`` because of this. This has
-the drawback of allowing contributions that introduce deprecations but:
+By default, any non-legacy-tagged or any non-`@-silenced`_ deprecation
+notices will make tests fail. Alternatively, you can configure an
+arbitrary threshold by setting ``SYMFONY_DEPRECATIONS_HELPER`` to
+``max[total]=320`` for instance.  It will make the tests fails only if a
+higher number of deprecation notices is reached (``0`` is the default
+value).
+
+You can even finer-grained control by using other keys of the ``max``
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]: Woops I forgot a word here:

```suggestion
You can have even finer-grained control by using other keys of the ``max``
```

File: components/phpunit_bridge.rst [hunk 7]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+array, which are ``internal``, ``direct``, and ``indirect``. The
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
array, which are ``self``, ``direct``, and ``indirect``. The
```

File: components/phpunit_bridge.rst [hunk 8]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+``SYMFONY_DEPRECATIONS_HELPER`` environment variable accept a
+url-encoded string, meaning you can combine thresholds and any other
+configuration setting, like this:
+``SYMFONY_DEPRECATIONS_HELPER=max[total]=42&max[internal]=0&verbose=0``
+
+Internal deprecations
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
Self deprecations
```

File: components/phpunit_bridge.rst [hunk 9]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+.....................
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
.................
```

File: components/phpunit_bridge.rst [hunk 10]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+but:
 
 * forget to fix the deprecated calls if there are any;
 * forget to mark appropriate tests with the ``@group legacy`` annotations.
 
-By using the ``"weak_vendors"`` value, deprecations that are triggered outside
-the ``vendors`` directory will make the test suite fail, while deprecations
-triggered from a library inside it will not, giving you the best of both
-worlds.
+By using ``SYMFONY_DEPRECATIONS_HELPER=max[internal]=0``,
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
By using ``SYMFONY_DEPRECATIONS_HELPER=max[self]=0``,
```

File: components/phpunit_bridge.rst [hunk 11]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+
+Direct and indirect deprecations
+................................
+
+When working on a project, you might be more interested in
+``max[direct]``. Let's say you want to fix deprecations as soon as
+they appear. A problem many people experience is that some dependencies
+they have tend to lag behind their own dependencies, meaning they do not
+fix deprecations as soon as possible, which means there is nothing you
+can do to fix those (apart from a pull request on the outdated vendor).
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]: @nicolas-grekas told me an issue he has with this is that it could make people less caring about their libs, so maybe rephrase that last part from "which means there is nothing you can do to fix those (apart from a pull request on the outdated vendor)", to "which means you should create a pull request on the outdated vendor, and ignore these deprecations until your pull request is merged"? Nicolas, I think you will probably still be unhappy about this, but this rewording will make you less unhappy, right?

File: components/phpunit_bridge.rst [hunk 12]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+keep up with the changes. You can of course still use ``max[indirect]``
+if you want to keep indirect deprecations under a given threshold.
+
+Here is a summary that should help you pick the right configuration:
+
++------------------------+-----------------------------------------------------+
+| Value                  | Recommended situation                               |
++========================+=====================================================+
+| max[total]=0           | Recommended for actively maintained projects        |
+|                        | with little to no dependencies                      |
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]: @nicolas-grekas told me about package principles, in particular, you are supposed to only have dependencies that are more stable than your package. Let's add a subtle hint about this.

```suggestion
|                        | with robust/no dependencies.                        |
```

File: components/phpunit_bridge.rst [hunk 13]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
++------------------------+-----------------------------------------------------+
+| max[internal]=0        | Recommended for libraries that use                  |
+|                        | the deprecation system themselves and               |
+|                        | cannot afford to use one of the modes above.        |
++------------------------+-----------------------------------------------------+
+
+Disabling the verbose output
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, the brige will display a detailed output with the number of
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
By default, the bridge will display a detailed output with the number of
```

File: components/phpunit_bridge.rst [hunk 14]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
+``max[total]=320`` for instance.  It will make the tests fails only if a
+higher number of deprecation notices is reached (``0`` is the default
+value).
+
+You can even finer-grained control by using other keys of the ``max``
+array, which are ``internal``, ``direct``, and ``indirect``. The
+``SYMFONY_DEPRECATIONS_HELPER`` environment variable accept a
+url-encoded string, meaning you can combine thresholds and any other
+configuration setting, like this:
+``SYMFONY_DEPRECATIONS_HELPER=max[total]=42&max[internal]=0&verbose=0``
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
``SYMFONY_DEPRECATIONS_HELPER=max[total]=42&max[self]=0&verbose=0``
```

File: components/phpunit_bridge.rst [hunk 15]
```diff
@@ -188,39 +188,89 @@ message contains the ``"foobar"`` string.
...
++------------------------+-----------------------------------------------------+
+| Value                  | Recommended situation                               |
++========================+=====================================================+
+| max[total]=0           | Recommended for actively maintained projects        |
+|                        | with little to no dependencies                      |
++------------------------+-----------------------------------------------------+
+| max[direct]=0          | Recommended for projects with dependencies          |
+|                        | that fail to keep up with new deprecations.         |
++------------------------+-----------------------------------------------------+
+| max[internal]=0        | Recommended for libraries that use                  |
```
Comment (greg0ire) [inline components/phpunit_bridge.rst]:
```suggestion
| max[self]=0            | Recommended for libraries that use                  |
```

**Review by greg0ire — CHANGES_REQUESTED** [review]
Hey! My PR is finally merged! :tada: Here is a review reflecting the latest changes.

**Review by greg0ire — APPROVED** [review]
Looks good!

File: components/phpunit_bridge.rst [hunk 16]
```diff
@@ -188,39 +188,90 @@ message contains the ``"foobar"`` string.
...
-the ``vendors`` directory will make the test suite fail, while deprecations
-triggered from a library inside it will not, giving you the best of both
-worlds.
+By using ``SYMFONY_DEPRECATIONS_HELPER=max[self]=0``,
+deprecations that are triggered outside the ``vendors`` directory will
+be accounted for seperately, while deprecations triggered from a library
+inside it will not (unless you reach 999999 of these), giving you
+the best of both worlds.
+
+Direct and indirect deprecations
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: Direct and Indirect Deprecations

File: components/phpunit_bridge.rst (hunk 8 above)
Comment (OskarStark) [reply components/phpunit_bridge.rst]: Internal Deprecations

File: components/phpunit_bridge.rst [hunk 17]
```diff
@@ -188,39 +188,90 @@ message contains the ``"foobar"`` string.
...
++------------------------+-----------------------------------------------------+
+| max[direct]=0          | Recommended for projects with dependencies          |
+|                        | that fail to keep up with new deprecations.         |
++------------------------+-----------------------------------------------------+
+| max[self]=0            | Recommended for libraries that use                  |
+|                        | the deprecation system themselves and               |
+|                        | cannot afford to use one of the modes above.        |
++------------------------+-----------------------------------------------------+
+
+Disabling the verbose output
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: Disabling the Verbose Output

File: components/phpunit_bridge.rst [hunk 18]
```diff
@@ -188,39 +188,90 @@ message contains the ``"foobar"`` string.
...
+Internal deprecations
+.....................
+
+When you maintain a library, having the test suite fail as soon as a
+dependency introduces a new deprecation is not desirable, because it
+shifts the burden of fixing that deprecation to any contributor that
+happens to submit a pull request shortly after a new vendor release is
+made with that deprecation. To mitigate this, you can either use tighter
+requirements, in the hope that dependencies will not introduce
+deprecations in a patch version, or even commit the Composer lock file,
```
Comment (OskarStark) [inline components/phpunit_bridge.rst]: composer.lock in double backticks

File: setup/upgrade_major.rst [hunk 19]
```diff
@@ -108,11 +108,11 @@ done!
             <!-- ... -->
 
             <php>
-                <env name="SYMFONY_DEPRECATIONS_HELPER" value="weak"/>
+                <env name="SYMFONY_DEPRECATIONS_HELPER" value="max[total]=999999"/>
             </php>
         </phpunit>
 
-    (you can also execute the command like ``SYMFONY_DEPRECATIONS_HELPER=weak phpunit``).
+    (you can also execute the command like ``SYMFONY_DEPRECATIONS_HELPER=max[total]=999999 phpunit``).
```
Comment (OskarStark) [inline setup/upgrade_major.rst]: I would prefer:

.. code-block:: terminal

    $ SYMFONY_DEPRECATIONS_HELPER=max[total]=999999 php phpunit

File: setup/upgrade_major.rst [hunk 20]
```diff
@@ -108,11 +108,15 @@ done!
...
+                <env name="SYMFONY_DEPRECATIONS_HELPER" value="max[total]=999999"/>
             </php>
         </phpunit>
 
-    (you can also execute the command like ``SYMFONY_DEPRECATIONS_HELPER=weak phpunit``).
+    You can also execute the command like:
+
+    .. code-block:: terminal
+
+        $ SYMFONY_DEPRECATIONS_HELPER=max[total]=999999 phpunit
```
Comment (OskarStark) [inline setup/upgrade_major.rst]: @javiereguiluz do we need to add „php“ in front of „phpunit“?

Comment (javiereguiluz) [reply setup/upgrade_major.rst]: In the docs, we don't use "pure PhpUnit" but PhpUnitBridge ... so we don't have examples of executing the regular PhpUnit script. However, thanks to the latest Symfony recipes, you don't have to do `./vendor/bin/simple-phpunit` because `./bin/phpunit` is equivalent (that's what we show in https://symfony.com/doc/current/testing.html for example).

Maybe we can use `./bin/phpunit` too here?

Comment (wouterj) [reply setup/upgrade_major.rst]: Then, we should use `php ./bin/phpunit` to have Windows compatibility

Comment (OskarStark) [reply setup/upgrade_major.rst]: Yes please

Comment (llaakkkk) [reply setup/upgrade_major.rst]: @OskarStark @wouterj done :)

**Review by OskarStark — APPROVED** [review]

**Comment by javiereguiluz** [comment]
@llaakkkk thanks a lot for this contribution! I'm happy to finally have it merged in Symfony Docs. Cheers!
